flow/layers/physical_shape_layer_unittests.cc

```diff
@@ -200,6 +200,34 @@ TEST_F(PhysicalShapeLayerTest, ElevationComplex) {
                0, MockCanvas::DrawPathData{layer_path, layer_paint}}}));
 }
 
+TEST_F(PhysicalShapeLayerTest, ShadowNotDependsCtm) {
+  constexpr SkScalar elevations[] = {1, 2, 3, 4, 5, 10};
```

Comment:
The original bugs seemed to be concerned about the look of the shadow as the size of the widget changed. It would be nice to test that.

Perhaps either some golden tests for shadows of various sizes of path, or a rendering test that draws shadows of 2 paths of different sizes, but with the top extending out of view so that the part visible in the generated image should look identical?

Reply:
That bug is not reproducible on latest version (https://github.com/flutter/flutter/issues/28463#issuecomment-762756193). But I can create a test with different sizes using `PhysicalShapeLayer::ComputeShadowBounds`. We can substract shadow bounds and path bounds, then checking them equal across different sizes (with same elevation). I don't want to add a golden test (Because I don't know).

Reply:
A test that rendered 2 shadows with paths of different sizes and expected the same size shadow at the bottom should be correct even if the values might eventually be tweaked for compliance with a spec. The point is how the shadow reacts to the path size and the answer should be "it shouldn't" if I am reading all of this right?

Something like:

Record a picture of a shadow of a rect LTRB(20, -50, 80, 80). Rasterize it to an image that is (0,0,100, >100) (large enough to capture the shadow below the rect, but not the top of the rect). Only the bottom of the rect and shadow should show.
Record the same thing, but the start Y of the rect is -500 or something larger, it still ends at Y=80 per the previous image.
The two images should be identical.

Do the same thing for the top of a small and large rect (have it extend off the bottom of the image this time).
Those two images should be identical.

The display_list_canvas_unittests.cc have code that renders a bunch of primitives and checks their output against each other that you can lift.

Even if we eventually tweak the values to match a spec, these tests still verify that the shadows above and below a path are unchanged by the size of the path.

Reply:
I reference this method in the display_list_canvas_unittests.cc. It's not testing whether the rendering is correct, but just that we render the same thing via direct calls to a surface canvas and indirectly through a DisplayList. If Skia fixes a bug that makes the rendering different, the same difference should appear in both images and still compare ==. Similarly, if we ever change how big or dark the shadow is to match a Material or Cupertino spec, we can still have a test that makes sure that the size of the "card list" or path doesn't affect the size or look of the shadow.

https://github.com/flutter/engine/blob/b57c5025159ba6f3c5703e2a6d2e433e4809bfcc/flow/display_list_canvas_unittests.cc#L666

Reply:
You can ignore the additional bounds testing in those cases, I was testing both the accuracy of the bounds from both mechanisms as well as the identical output. I was also testing that "something is rendered" (i.e. pixels_touched > 0) because I wanted to make sure that none of my tests requested a NOP rendering, but that isn't a critical part of the test.

Reply:
Thanks for detailed tutorial @flar. I created the test.

Reply:
NP, we should probably add a test utility to do things like this, but don't worry about that for now. I'll take that as an action item.